### PR TITLE
test: tx replay handle forks 

### LINKED ISSUE #5971: Tx Replay: handle forks that occur during transaction replay
If a fork occurs, which causes signers to enforce transaction replay, it's possible for another fork to occur in the middle of a replay. When this occurs, signers must enforce transaction replay on the entire "in-progress" replay as if it was starting from scratch.

### PR DESCRIPTION
**Pull request by fdefelici**
### Description

This PR implements the logic to allow Tx Replay to respond to forks while it is in progress.

The relevant tests are those prefixed with tx_replay_with_fork_*. In particular, I want to highlight the following two tests, which demonstrate how an ongoing replay is updated:
- `tx_replay_with_fork_causing_replay_set_to_be_updated`: updates the replay set from the deepest fork.
- `tx_replay_with_fork_causing_replay_to_be_cleared_due_to_cycle`: clears the replay set when the deepest fork begins from a previous reward cycle.

To handle these scenarios, I needed to introduce a new state to track the burn blocks - fork origin and tip - as they were at the time the Tx Replay started.

Currently, this has been implemented as part of the local signer state to demonstrate functionality while keeping the impact on the code-base minimal:
```rust
pub struct Signer {
    ...    
    pub tx_replay_scope: TxReplayScopeOpt,
}
```

However, it may be more appropriate to include this information as part of the State Machine update (which would affect the signer agreement), since this state logically follows the behavior of `SignerStateMachine::tx_replay_set`.

> If we go this route, it might make sense to introduce a new data structure, TxReplayInfo, to represent two distinct states:
> - "Empty": no Tx Replay is active
> - "In Progress": a replay is ongoing, containing both the replay set and its scope

If that’s the preferred direction, I’d be happy to refactor the solution and move the state into the `SignerStateMachine` and related components.

### Applicable issues

- fixes #5971 

### Additional info (benefits, drawbacks, caveats)

Alongside the fork-related tests, this PR also adds a couple of tests for edge cases:
- `tx_replay_solved_by_mempool_txs`: demostrating that Tx Replay can be "solved" by chance picking txs from the mempool (and not from the signer replay set), and
- `tx_replay_rejected_for_invalid_proposal_due_to_partial_replay` to illustrate a potential weakness in the replay validation mechanism (currently failing on purpose)
  - The scenario simulates a miner submitting a block that only includes txA, while the replay set is [txA, txB].
  - This is done using a test flag to control block construction.
  - Currently, the block is incorrectly considered valid, even though there's no valid reason (e.g. Tenure Budget exhaustion) for excluding txB.
  - Ideally, the block should be rejected by the signers, as the miner could have included both transactions.
  - This test serves as a starting point for proper reasoning and eventually be handled in a dedicated PR (potentially related to #5972). ***However let me know if it could be useful, or it is better to remove it***

NOTE: `tx_replay_rejected_for_invalid_proposal_due_to_partial_replay` removed in 4874e7f034fa96ad3eb0b8d13ee86bd1d26a4b1d, because we want to allow this behaviour.

### Checklist

- [x] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

### CODE REVIEW

File: stacks-signer/src/v0/signer_state.rs [hunk 1]
```diff
@@ -88,6 +88,11 @@ pub struct NewBurnBlock {
     pub consensus_hash: ConsensusHash,
 }
 
+/// Scope of Tx Replay in terms of Burn block boundaries
+/// - tuple.0 is the fork originating the tx replay,
+/// - tuple.1 is the canonical burnchain tip when Tx Replay begun
+pub type TxReplayScopeOpt = Option<(NewBurnBlock, NewBurnBlock)>;
```
Comment (hstove) [inline stacks-signer/src/v0/signer_state.rs]: Can we make this a struct? Structs just generally feel easier to read

File: stacks-signer/src/v0/signer_state.rs [hunk 2]
```diff
@@ -915,10 +929,52 @@ impl LocalStateMachine {
...
+                "expected_burn_block.hash" => %expected_burn_block.consensus_hash,
+                "prior_state_machine.burn_block_height" => prior_state_machine.burn_block_height,
+                "prior_state_machine.burn_block" => %prior_state_machine.burn_block,
+            );
+
+            let (fork_origin, past_tip) = match tx_replay_scope {
+                Some(scope) => scope,
+                None => {
+                    warn!("Tx Replay: BUG! Scope cannot be None while in replay mode!");
+                    return Err(SignerChainstateError::LocalStateMachineNotReady);
```
Comment (hstove) [inline stacks-signer/src/v0/signer_state.rs]: I think this is a good example of what you mentioned about combining this new state data structure with `ReplayTransactionSet` (at least for the `LocalStateMachine`) so that we can type-guard this situation. Then, `handle_possible_bitcoin_fork` can return this combined data structure

File: stacks-signer/src/v0/signer_state.rs [hunk 3]
```diff
@@ -936,37 +992,89 @@ impl LocalStateMachine {
...
+        let target_reward_cycle = reward_cycle_info.get_reward_cycle(fork_tip.burn_block_height);
         let is_fork_in_current_reward_cycle = fork_info.iter().all(|fork_info| {
             let block_height = fork_info.burn_block_height;
             let block_rc = reward_cycle_info.get_reward_cycle(block_height);
-            block_rc == current_reward_cycle
+            block_rc == target_reward_cycle
         });
+
         if !is_fork_in_current_reward_cycle {
-            info!("Detected bitcoin fork occurred in previous reward cycle. Tx replay won't be executed");
```
Comment (hstove) [inline stacks-signer/src/v0/signer_state.rs:969]: Can we keep this log, or a similar log, here?

File: stacks-signer/src/v0/signer_state.rs [hunk 4]
```diff
@@ -936,37 +992,89 @@ impl LocalStateMachine {
...
+            expected_burn_block,
+            &potential_replay_tip,
+        )? {
+            None => {
+                info!("Detected bitcoin fork occurred in previous reward cycle. Tx replay won't be executed");
+                Ok(None)
+            }
+            Some(replay_set) => {
+                let scope_opt = if !replay_set.is_empty() {
+                    let scope = (expected_burn_block.clone(), potential_replay_tip);
```
Comment (hstove) [inline stacks-signer/src/v0/signer_state.rs]: In this branch, you have this check (`if !replay_set.is_empty()`), but not in the other branch (where we're already in a replay). IMO this is another argument for a combined data structure, and probably to just return that combined data structure in `compute_forked_txs_set_in_same_cycle`

File: stackslib/src/chainstate/stacks/miner.rs [hunk 5]
```diff
@@ -3005,6 +3027,13 @@ fn select_and_apply_transactions_from_vec<B: BlockBuilder>(
     debug!("Replay block transaction selection begins (parent height = {tip_height})");
     for replay_tx in replay_transactions {
         fault_injection_stall_tx();
+        #[cfg(any(test, feature = "testing"))]
+        {
+            if should_skip_replay_tx(replay_tx.txid()) {
+                continue;
+            }
+        }
```
Comment (hstove) [inline stackslib/src/chainstate/stacks/miner.rs]: nit, but I think it's a little better (and more consistent) to follow the pattern of the line above this (`fault_injection_stall_tx()`), where the `cfg` macro is instead around the fault injection function

File: stacks-signer/src/v0/signer_state.rs (hunk 3 above)
Comment (fdefelici) [reply stacks-signer/src/v0/signer_state.rs:969]: That log has been moved to the `compute_forked_txs_set_in_same_cycle` callers, and more specifically where the `None` branch is managed.

This way the `compute_forked_txs_set_in_same_cycle` contains mere calculation and the callers can apply proper logging.

Do you think is not enough?

File: stackslib/src/chainstate/stacks/miner.rs (hunk 5 above)
Comment (fdefelici) [reply stackslib/src/chainstate/stacks/miner.rs]: I thought about that at first, but in this case the objective is not to "stall" the loop, but skipping the listed transactions.

So, the cleanest solution I came up with was this.

Have you any hint, on how to do that keeping the "skipping" logic?

File: stacks-signer/src/v0/signer_state.rs (hunk 1 above)
Comment (fdefelici) [reply stacks-signer/src/v0/signer_state.rs]: Yes sure! I didn't because waiting for decision if keeping this as a `Signer` "local" state or move it to the `SignerStateMachine` creating a brand new data structure involing also the `tx_replay_set`

File: stacks-signer/src/v0/signer_state.rs (hunk 4 above)
Comment (fdefelici) [reply stacks-signer/src/v0/signer_state.rs]: Yes, definitly (currently not needed this check when we are already in replay because of the `is_in_tx_replay_mode` boolean)

File: stacks-signer/src/v0/signer_state.rs (hunk 2 above)
Comment (fdefelici) [reply stacks-signer/src/v0/signer_state.rs]: Yes, exactly!

If we're aligned on moving all this logic into the `SignerStateMachine` (within `LocalStateMachine`), I’ll go ahead and start refactoring. The plan would be to introduce a new data structure that encapsulates both `TxReplayScope` and `ReplayTransactionSet`.

Shall I proceed?

Comment (hstove) [reply stacks-signer/src/v0/signer_state.rs]: Hey - yes, I'd say go for it (in regards to updating the data structure)

File: stacks-signer/src/v0/signer_state.rs (hunk 3 above)
Comment (hstove) [reply stacks-signer/src/v0/signer_state.rs:969]: yep you're right, i missed that 👍

File: stackslib/src/chainstate/stacks/miner.rs (hunk 5 above)
Comment (hstove) [reply stackslib/src/chainstate/stacks/miner.rs]: I think the idea would be to just have:

```
if should_skip_replay_tx(replay_tx.txid()) {
    continue;
}
```

So, just removing the `[cfg(test)]` block, and instead having that macro at the definition of `should_skip_tx_replay`

File: stacks-signer/src/v0/signer_state.rs (hunk 4 above)
Comment (hstove) [reply stacks-signer/src/v0/signer_state.rs]: Ah, true!

File: stackslib/src/chainstate/stacks/miner.rs (hunk 5 above)
Comment (fdefelici) [reply stackslib/src/chainstate/stacks/miner.rs]: Ok. I didn't initially consider this since it's not a one-liner like `fault_injection_stall_tx()`.
That said, if this is the preferred approach, I update it accordingly.

Comment (fdefelici) [reply stackslib/src/chainstate/stacks/miner.rs]: Done in efe0f603fae53c7dc01c4ca0d6c664a76421ce9c

File: stacks-signer/src/v0/signer_state.rs (hunk 2 above)
Comment (fdefelici) [reply stacks-signer/src/v0/signer_state.rs]: Approached the refactoring in two ways:
- ReplayState as an enum: https://github.com/fdefelici/stacks-core/pull/1
- ReplaState as a struct: https://github.com/fdefelici/stacks-core/pull/2

Each comes with its own trade-offs:
- `Enum-based`: Offers a stricter, more type-safe design, but requires exhaustive match handling, which can be verbose.
- `Struct-based`: Provides more flexibility and is easier to work with in branching logic, but lacks the compiler-enforced guarantees that enums provide.

Both PRs are opened locally against this branch, so the scope of changes introduced by the refactor is fully contained.

Let me know which direction you prefer, and I can move forward accordingly.

Comment (hstove) [reply stacks-signer/src/v0/signer_state.rs]: Thanks for putting together these two options! I think we should go with the Enum-based pattern.

Comment (fdefelici) [reply stacks-signer/src/v0/signer_state.rs]: Applied Enum-based solution in 8950112b7d98d8eeae63b1762dfd4db902d96475

File: stacks-signer/src/v0/signer_state.rs (hunk 1 above)
Comment (fdefelici) [reply stacks-signer/src/v0/signer_state.rs]: this was done here: 5c6780cc3687e19bfb7455709fd3f47f9b3fc5dd

File: stacks-signer/src/v0/signer_state.rs [hunk 6]
```diff
@@ -88,10 +88,48 @@ pub struct NewBurnBlock {
...
+    /// - `Unset` if the replay set is empty.
+    /// - `Invalid` if the replay set exists but the scope is missing.
+    /// - `InProgress` if both a non-empty replay set and a valid scope are provided.
+    fn infer_state(replay_set: &ReplayTransactionSet, scope_opt: &ReplayScopeOpt) -> Self {
+        if replay_set.is_empty() {
+            return Self::Unset;
+        }
+
+        match scope_opt {
+            None => Self::Invalid,
```
Comment (hstove) [inline stacks-signer/src/v0/signer_state.rs]: I think we should just return `Unset` here, along with an `error!` log. Using `Unset` simplifies the code, and we'd want to put a log message here regardless IMO.

Comment (hstove) [reply stacks-signer/src/v0/signer_state.rs]: And if we did that, I think we should just remove the `Invalid` enum.

Comment (hstove) [reply stacks-signer/src/v0/signer_state.rs]: Hm... now looking at the code below this, I'm not totally sure. Let's sync on this at standup.

Comment (fdefelici) [reply stacks-signer/src/v0/signer_state.rs]: The `Invalid` variant currently represents a case where the Signer's local state (scope) is not aligned with the ongoing Replay Set.

I implemented it like this, to give a "chance" to the Signer, in case ReplayState is "invalid" but `handle_possible_bitcoin_fork(..)` could return `Ok(None)` (e.g. no fork detected).

However, if we can be more "strict" (about the invalid replay state), we could consider simplifying this by:
- Removing the `Invalid` variant, and 
- Computing the `ReplayState` as an `Option`:
  - If it’s `None`, it indicates the state is "invalid", and we can return a LocalStateMachine error right away,
  - Otherwise, it’s Some, we proceed with the valid `ReplayState` in `handle_possible_bitcoin_fork(..)`

Comment (fdefelici) [reply stacks-signer/src/v0/signer_state.rs]: Tried the above approach here: 686894a0a46e2079e9d6b5530965a8228a6e0c33

Let me know if this could be preferred, or if I have to rollback to the previous approach

**Review by hstove — APPROVED** [review]

**Review by kantai — APPROVED** [review]
